### PR TITLE
Porting math.js blocks to JSON

### PR DESCRIPTION
**Pull request by AnmAtAnm**
Moving all `math.js` definitions into a single JSON array, complete with i18n syntax for all messages, dropdowns, and tooltips. Mutators and dynamic tooltips have been ported as extensions.

Adding `Blockly.Extensions.buildTooltipForDropdown(..)` to facilitate the creation and error-checking of tooltips that update based on the value of a dropdown.

`FieldDropdown.getOptions()` is now public (was `.getOptions_()`). I also added `FieldDropdown.isOptionListDynamic()`.

Now warns (and corrects) when `jsonInit()` encounters a raw string in the 'extensions' attribute, where an array is expected.

### CODE REVIEW

File: core/extensions.js [hunk 1]
```diff
@@ -75,6 +75,72 @@ Blockly.Extensions.apply = function(name, block) {
...
+ * Builds an extension function that will map a dropdown value to a tooltip string.
+ * Tooltip strings will be passed through Blockly.utils.checkMessageReferences(..)
+ * immediately and Blockly.utils.replaceMessageReferences(..) at display time.
+ * @param {string} dropdownName The name of the field whose value is the key
+ *     to the lookup table.
+ * @param {!Object<string, string>} lookupTable The table of field values to
+ *     tooltip text.
+ * @return {Function} The extension function.
+ */
+Blockly.Extensions.buildTooltipForDropdown = function(dropdownName, lookupTable) {
```
Comment (rachel-fenichel) [inline core/extensions.js]: This function has too many nested functions inside it.  Please decompose it.

File: core/extensions.js [hunk 2]
```diff
@@ -75,6 +75,72 @@ Blockly.Extensions.apply = function(name, block) {
...
+
+      blockTypesChecked.push(this.type);
+    }
+
+    this.setTooltip(function() {
+      var value = thisBlock.getFieldValue(dropdownName);
+      var tooltip = lookupTable[value];
+      if (tooltip == null) {
+        if (!blockTypesChecked.includes(thisBlock.type)) {
+          // Warn for missing values on gnereated tooltips
```
Comment (rachel-fenichel) [inline core/extensions.js]: *generated

File: core/extensions.js (hunk 1 above)
Comment (AnmAtAnm) [inline core/extensions.js]: I extracted `checkDropdownOptionsInTable_(..)`, as suggested.  Same function depth.  :/

File: core/extensions.js [hunk 3]
```diff
@@ -75,6 +75,85 @@ Blockly.Extensions.apply = function(name, block) {
...
+    // Validate all dropdown options have values.
+    var dropdown = block.getField(dropdownName);
+    if (!dropdown.isOptionListDynamic()) {
+      var options = dropdown.getOptions();
+      for (var i = 0; i < options.length; ++i) {
+        var optionKey = options[i][1]; // label, then value
+        if (lookupTable[optionKey] == null) {
+          console.warn('No tooltip mapping for value ' + optionKey +
+            ' of field ' + dropdownName + ' of block type ' + block.type);
+        } else {
```
Comment (rachel-fenichel) [inline core/extensions.js]: Empty else.

File: core/extensions.js [hunk 4]
```diff
@@ -75,6 +75,85 @@ Blockly.Extensions.apply = function(name, block) {
...
+        this, dropdownName, lookupTable);
+      blockTypesChecked.push(this.type);
+    }
+
+    this.setTooltip(function() {
+      var value = thisBlock.getFieldValue(dropdownName);
+      var tooltip = lookupTable[value];
+      if (tooltip == null) {
+        if (!blockTypesChecked.includes(thisBlock.type)) {
+          // Warn for missing values on generated tooltips
```
Comment (rachel-fenichel) [inline core/extensions.js]: Extract here through 121 as a helper function.

Comment (AnmAtAnm) [inline core/extensions.js]: A helper function to print a warning, largely concatenating strings seems excessive.  Collapsed one line.

**Pull request merged**